funkin: init at unstable-2021-04-02

OPNA2608 opened the pull request:
*getting freaky on a friday night yeah*

![Bildschirmfoto von 2021-03-19 00-04-18](https://user-images.githubusercontent.com/23431373/111708732-b0c94200-8846-11eb-8283-484c72b3d2cb.png)

###### Motivation for this change
Packaging the game [Friday Night Funkin'](https://ninja-muffin24.itch.io/funkin). May require some tuning of your PulseAudio device's latency settings to feel comfortable, certainly did for me.

~~As noted in the file the source code is available at https://github.com/ninjamuffin99/Funkin but requires new haxePackages entries & a newer version of Haxe itself (#116953 perhaps, although it needs 4.1.x specifically) to compile. For that reason I've only packaged the prebuilt 64-bit Linux build from the itch.io page for now. I'll give a source build a shot if that PR gets merged, though I've never done anything with Haxe before.~~

We are now building the game from source. Only `x86_64-linux` (any maybe `i686-linux`, untested) compile with our current way of packaging Haxe packages.

The game complains on launch about an updated version being available, ~~but that elusive v0.2.7.1 version lacks a Linux build on itch.io so *¯\\\_(ツ)\_/¯*.~~ that's a currently-broken version check in the game. Pressing space will also fail to bring up the itch.io page because of a hardcoded `/usr/bin/xdg-open`, I've mentioned that problem on a relevant issue of the github repo. Adding a FSH env just to get that to work feels overkill so I didn't bother with it. (Not that I'd know how to make a FSH env for a package either)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 116800` at aaeae72e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build (new failures):</summary>
  <ul>
    <li><a href="https://gist.github.com/c4339117cae246034fef21299807b414">funkin</a></li></li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/games/funkin/default.nix:
```diff
@@ -0,0 +1,104 @@
+    homepage = "https://ninja-muffin24.itch.io/funkin";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ OPNA2608 ];
+    platforms = with platforms; windows ++ darwin ++ linux;
+    broken = stdenv.system != "x86_64-linux";
```
```suggestion
    platforms = with platforms; [ "x86_64-linux" ];
```

OPNA2608 replied:
I guess that's fine while we're not building from source.

OPNA2608 replied:
I've re-added the original `broken` line because this should be buildable from source on more than just x86_64-linux, but everything else either has binary format or linking problems because of pre-compiled executables and libraries in some Haxelib packages. I've added a comment explaining the situation.

sternenseemann commented:
I can still do haxe 4.1, but I'd prefer to do it in a separate PR. Any chance you could help me with testing haxe 4.2? Like anything to try and build. I want to rule out any hard coded paths left over or similar before merging.

OPNA2608 commented:
I think I can get this to build with 4.2 actually, it's just a matter of building various dependencies from github commits instead of their latest release.

OPNA2608 commented:
4.2 seems to suffice for building both the latest release & current master. I'll start redoing this PR when I have more time, prolly \~ Easter holidays.

OPNA2608 committed:
haxePackages.hxcpp: 4.1.15 -> 4.2.1

OPNA2608 committed:
haxePackages.hscript: init at 2.4.0

OPNA2608 committed:
haxePackages.lime: init at 7.9.0

OPNA2608 committed:
haxePackages.openfl: init at 9.1.0

OPNA2608 committed:
haxePackages.flixel: init at 4.8.1

OPNA2608 committed:
haxePackages.flixel-addons: init at 2.9.0

OPNA2608 committed:
haxePackages.flixel-ui: init at 2.3.3

OPNA2608 committed:
haxePackages.newgrounds: init at 1.1.4

OPNA2608 committed:
haxePackages.polymod: init at unstable-2021-04-06

OPNA2608 committed:
haxePackages.discord_rpc: init at unstable-2021-03-26

marius851000 commented:
I was able to compile the game from source with no majot issue. I just need to :
1. add the desktop icon
2. unbundle lime

OPNA2608 commented:
@marius851000 I have a local revision of this branch that builds everything from source, just waiting for #118112 so I don't have to bump any Haxelib packages to git-only versions. I can push them to this PR this evening if you want to take a look at it.

sternenseemann commented:
Sorry, I forgot you were blocking on my PR. I'll clean it up and merge it later.

OPNA2608 commented:
@marius851000 I pushed what I got, might need more work to get pretty but it's a start. Hardwires Linux for now, haven't tried on Darwin yet. I've packaged a git commit here, but 0.2.8 is slowly being released so I'll bump the derivation to that once it's on GitHub.

Like you mentioned, `lime` is sorta ugly. I think it includes prebuilt unpatched libraries, I'm unsure if it just copies the appropriate one into the build directory of whatever project depends on `lime` or if it's rebuilt in any way.

I feel like one (or more) of the Haxelib packages added here should propagate the packages that the final game depends on in its wrapping, but I haven't put much thought into that yet. I feel like the entire Haxe infrastructure could use some work in general but I have enough projects to worry about for now :sweat_smile:.

OPNA2608 commented on pkgs/games/funkin/default.nix:
```diff
@@ -0,0 +1,105 @@
+    runHook preBuild
+
+    export HOME=$PWD
+    export HXCPP_COMPILE_THREADS=${if enableParallelBuilding then "$NIX_BUILD_CORES" else "1"}
```
Getting this propagated through the `hxcpp` package would be great. Without this the compile process uses as many threads as your CPU has in total, which is very unfriendly for obvious reasons, and it took me a good while to find this environment variable.

sternenseemann replied:
This should be pretty easy to do by adding a `setupHook` to the haxe* derivations.

sternenseemann replied:
Err, hxcpp of course

OPNA2608 commented:
> Hardwires Linux for now, haven't tried on Darwin yet.

I gave this a quick attempt. After the PRs that fix the haxe package itself and excluding all the Linux-specific runtime dependencies, running `haxelib run lime build mac -final` yields only this response:

```
Could not link to neko.
Could not link plugin to process (CFFILoader.h 269) - with neko
```

These messages apparently result from [hxcpp's cffi loader being unable to dlopen `neko` library](https://github.com/HaxeFoundation/hxcpp/blob/e23515d4acd43f3609e5ca49a8cebdad285db78d/include/hx/CFFINekoLoader.h#L23-L43). Haven't looked too far into this yet.

OPNA2608 commented:
> I've packaged a git commit here, but 0.2.8 is slowly being released so I'll bump the derivation to that once it's on GitHub.

[0.2.8 won't be released publicly for a while](https://twitter.com/ninja_muffin99/status/1385290523955585026). I'll clean this up & mark it ready for review tonight, still x86_64-linux only because that's the only platform where all our haxelib binary packages seem to work on.

OPNA2608 committed:
haxePackages: Allow passing src instead of sha256

For when the source should be fetched from somewhere other than the
Haxe Library (e.g. GitHub).

OPNA2608 committed:
funkin: init at unstable-2021-04-02

OPNA2608 commented:
I'll close this for now & maybe pick this up again when a new proper version releases and I feel motivated to play this game again.

marius851000 commented:
Do you mind if I open a new PR with this (and my change for the desktop) ? I've been playing with making a nix user repository that could build fork of FNF, and I need those haxe dependancies.

marius851000 commented:
Without any response, I'll open a new PR with my small desktop change (and double-checking there is no issue too) later today.
